### PR TITLE
Nilability checking improvements #4

### PR DESCRIPTION
**Pull request by vasslitvinov**
Get a bunch more tests to pass with `--no-legacy-nilable-classes`,
with fixes to compiler and modules.

A couple of things of note.

* This fixes the compiler not to complain about "Cannot default-initialize"
  for params - they will get a "not of a supported param type" anyway.

* In `SSCA2`, I would replace lots of `Active_Level[here.id]` with a const.
  However that would be considered a style change and so would necessitate
  review by a wider subset of the group. This is more than I am shooting for
  in this PR. Also, some `Active_Level[here.id]` would not be replaceable
  as elegantly due to differing `here`, so I'd leave them as-is.
  At which point the replacement loses its appeal of elegance to me.

* In `Sort.chpl`, the helper functions `chpl_check_comparator()` and
  `radixSortOk()` play it loose with default-initalizing non-nilable
  variables. Given that we probably want those variables to have
  non-nilable types and that these functions are `param` so initializing
  them wouldn't help, I just marked them with `pragma "unsafe"`.
  Makes me wonder if all `param` functions should/could be considered
  "unsafe" in this regard.

* In AMR / `BergerRigoutsosClustering.chpl`, I have a better fix
  to keep the `signatures` field non-nilable. I do so by making a recursive
  function that computes the desired tuple and stores it into the field.
  This fix hits a limitation of `_createFieldDefaults()` so can't go in
  right now. I have a fix for that limitation, after which I can come back
  and revisit `signatures`.

* Wow getting AMR to pass the checks really took some work!

* There is an interesting pattern in `AMRHierarchy_def.chpl` :

```chpl
var refined_coarse = <something non-nilable>;
... lots of method calls on refined_coarse ...

var unfilled_intersection = <something else non-nilable>;
... lots of method calls on unfilled_intersection ...
delete unfilled_intersection;  unfilled_intersection = nil;

... lots more method calls on refined_coarse ...
delete unfilled_region;  unfilled_region = nil;
```

We surely want these two variables to be non-nilable, to avoid
the `!` operators every time a method is invoked on one of them.
This unfortunately conflicts with our desire to set them to `nil`
once we are done with them, to ensure they are not used afterwards.
I point this out as a familiar C-like idiom that is not smooth here.
Of course, the user should not be using `unmanaged` pointers. :)

Testing:
- [x] linux64 -futures

### CODE REVIEW

File: modules/dists/PrivateDist.chpl [hunk 1]
```diff
@@ -147,7 +147,7 @@ class PrivateDom: BaseRectangularDom {
 
 class PrivateArr: BaseRectangularArr {
   var dom: unmanaged PrivateDom(rank, idxType, stridable);
-  var data: eltType;
+  var data: eltType?;
```
Comment (mppf) [inline modules/dists/PrivateDist.chpl]: This does not seem right to me. Doesn't it mean that accessing a PrivateArr (via something dmapped Private) will give you back a nilable element even if you asked for a non-nilable element type?

I would expect code using PrivateDist with classes to use a nilable type when creating the Private array of classes.

File: spec/Classes.tex [hunk 2]
```diff
@@ -1566,8 +1566,8 @@ \subsection{Class Assignment}
 \end{chapelpre}
 \begin{chapel}
 // assume that C is a class
-var a = new owned C();
-var b:owned C; // default initialized to store `nil`
+var a:owned C? = new owned C();
```
Comment (mppf) [inline spec/Classes.tex]: I would expect it to continue to have `var a = new owned C();`. Does it not work that way today?

File: test/library/packages/UnitTest/PopulateTests_PRIM/TestCount.compopts [hunk 3]
```diff
@@ -0,0 +1 @@
+--legacy-nilable-classes
```
Comment (mppf) [inline test/library/packages/UnitTest/PopulateTests_PRIM/TestCount.compopts]: For other UnitTest tests, I've been adding `borrowed Test` to the arguments of the tests. That is, assuming you are seeing an error about unable to make a generic FCF.

File: test/library/packages/VectorList/VectorLists.chpl [hunk 4]
```diff
@@ -142,8 +142,10 @@ module VectorLists {
   /* Yields all VectorListElement. */
   inline iter VectorList.listElements() {
     var curr = head;
-    do { const next = curr.next; yield curr; curr = next; }
-    while curr != nil;
+    while true {
+      const next = curr.next; yield curr;
```
Comment (mppf) [inline test/library/packages/VectorList/VectorLists.chpl]: This is test code so I'm not too picky but I don't like these multiple statements per line and think each `;` statement should get its own line.

File: modules/dists/PrivateDist.chpl (hunk 1 above)
Comment (vasslitvinov) [reply modules/dists/PrivateDist.chpl]: Good point. I adjusted.

File: spec/Classes.tex (hunk 2 above)
Comment (vasslitvinov) [reply spec/Classes.tex]: `a` is emptied two lines down then accessed after being emptied.

File: test/library/packages/UnitTest/PopulateTests_PRIM/TestCount.compopts (hunk 3 above)
Comment (vasslitvinov) [reply test/library/packages/UnitTest/PopulateTests_PRIM/TestCount.compopts]: PRIM_GATHER_TESTS does not work with --no-legacy-nilable-classes due to #13720.

**Pull request merged**